Add `map_word` and `parabolic_subgroup` for Weyl groups

TWiedemann opened the pull request:
Adds `parabolic_subgroup` as discussed in #4483. This relies on the new `map_word`. I also added `parabolic_subgroup_with_projection` (maybe the name is not good) which, in addition, returns the projection homomorphism from the large Weyl group to the parabolic subgroup.

I set this to draft because:
1. Someone will have to tell me where precisely to put the exports. I assume they go to experimental/LieAlgebras/src/exports.jl. For the moment I put them there at the end of file (where I will sort them alphabetically once this is clarified).
2. Examples/jldoctests for parabolic_subgroup(_with_projection) are still missing.

TWiedemann committed:
Add map_word and parabolic_subgroup(_with_projection) and corresponding exports

TWiedemann commented:
In my function `map_word(::WeylGroupElem, ...)`, I had to call `map_word(::Vector{Int}, ...)` with `Oscar.map_word` while `map_word(::Vector{Int}, ...)` worked fine on the REPL. Not sure what is going on there. Probably related to the doctest failure:

```
WARNING: using LieAlgebras.map_word in module Oscar conflicts with an existing identifier.
[...]
ERROR: MethodError: no method matching map_word(::WeylGroupElem, ::Vector{Int64})
```

thofma commented:
You need to add `map_word` here: https://github.com/TWiedemann/Oscar.jl/blob/tw/weyl_parabolics/experimental/LieAlgebras/src/LieAlgebras.jl#L72

fingolfin commented on experimental/LieAlgebras/src/exports.jl:
```diff
@@ -51,3 +51,7 @@ export tensor_power
+
+export parabolic_subgroup
+export parabolic_subgroup_with_projection
+export map_word
```
This file is the right place for exports. But  you don't need to `export map_word` here, instead you need to make sure to `import` it from outside (@thofma already pointed out the place were to do it). Then the `map_word` method you define will be a method for `Oscar.map_word`, instead of defining a new unrelated function `Oscar.LieAlgebras.map_word` (more precisely, if you `import map_word` from Oscar then `Oscar.map_word` will be identical to `Oscar.LieAlgebras.map_word`)

TWiedemann replied:
Done

fingolfin commented on experimental/LieAlgebras/test/WeylGroup-test.jl:
```diff
@@ -135,4 +135,97 @@
+      end
+    end
+  end
+end
```
Please add a trailing line

TWiedemann replied:
Done

fingolfin commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+@doc raw"""
+    map_word(w::WeylGroupElem, genimgs::Vector; init = nothing)
+
+If `init` is `nothing` and `word(w) = [UInt(`$i_1$`), ..., UInt(`$i_n$`)]`, then return the product $R_1 R_2 \cdots R_n$
```
I would omit the `UInt` for clarity
```suggestion
If `init` is `nothing` and `word(w) = [`$i_1$`, ..., `$i_n$`]`, then return the product $R_1 R_2 \cdots R_n$
```

TWiedemann replied:
I thought about that as well, but the trade-off between clarity and correctness wasn't evident for me. Anyway, I'll remove it.

fingolfin commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+See also: [`map_word(::Union{FPGroupElem, SubFPGroupElem}, ::Vector)`](@ref),
+[`map_word(::Union{PcGroupElem, SubPcGroupElem}, ::Vector)`](@ref).
+Note that `map_word(::WeylGroupElem)` does not provide a `genimgs_inv` parameter
+because the generators of a Weyl group are always self-inverse.
```
I think this is a mistake, as it precludes writing generic code.

Instead I would accept it anyway, but you can e.g. ignore it:
```suggestion
Note that `map_word(::WeylGroupElem)` accepts but ignores the `genimgs_inv`
keyword argument because the generators of a Weyl group are always self-inverse.
```
(One could also check if `genimgs == genimgs_inv` if both are given, but I wouldn't bother)

TWiedemann replied:
Makes sense! Done

fingolfin commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+@doc raw"""
+    parabolic_subgroup(W::WeylGroup, vec::Vector{<:Integer}, w::WeylGroupElem) -> WeylGroup, Map{WeylGroup, WeylGroup}
+
+Returns two objects: The subgroup `U` of `W` generated by `[inv(w)*u*w for u in gens(W)[vec]]`, whose generators are set to be this vector and which is itself a `WeylGroup`, and the embedding of `U` into `W`. The elements of `vec` must be pairwise distinct integers in `1:number_of_generators(W)` and `vec` must be non-empty.
```
Please wrap lines at something between 80 and 100 columns wide (it helps with reviewing changes made to this text in the future, as the diffs are nicer)

fingolfin replied:
- it should be "Return" not "Returns" (see point 2 at <https://docs.julialang.org/en/v1/manual/documentation/>)
- I'd say the return group is *not* that subgroup `U`, rather it is a free standing Weyl group isomorphic to `U`, and return map is an isomorphism between it and `U`.

How about this:
```suggestion
Return a Weyl group `P` and an embedding $f:P\to W$ such that $f(P)$ is
the subgroup `U` of `W` generated by `[inv(w)*u*w for u in gens(W)[vec]]`.
The elements of `vec` must be pairwise distinct integers in
`1:number_of_generators(W)` and `vec` must be non-empty.
```



Oh and an example in form of `jldoctest` would be super nice :-)

fingolfin replied:
Probably should also say that `f` maps the i-th generator of P t the i-th generator of `U` or so (at least I assume that's what it does?)

TWiedemann replied:
Done. The statement that $f(P)$ is generated by `[inv(w)*u*w for u in gens(W)[vec]]` is now technically redundant because it follows from the description of `f` on the generators, but I left it in because it seems more clear to me that way.

lgoettgens commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+Otherwise return the product $xR_1 R_2 \cdots R_n$ where $x =$ `init`.
+
+The length of `genimgs` must be equal to the rank of the parent of `w`.
+In particular, it must not be empty.
```
These two sentences imply that `map_word` would not work for a rank 0 Weyl group. 
Could you please make sure that it still works in this case by just returning `init` (in case that is set). This would make it analogous to `map_word(one(free_group), []; ...)`

TWiedemann replied:
Are rank 0 weyl groups supported? I cannot find a way to create one, which is why I wrote it this way. (Anyway, no harm in returning init in this case anyway.)

lgoettgens replied:
```
 weyl_group(zero_matrix(ZZ, 0, 0))
```
I think this is not advertised anywhere, but useful in tests to catch some edge-cases

TWiedemann replied:
I see. I'll change the documentation. The code can stay the same because `map_word` for integer vectors has the desired behaviour.

lgoettgens commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+julia> map_word(W([1]), imgs)
+2
+
+julia> map_word(W([1]), imgs, init=7)
```
```suggestion
julia> map_word(W([1]), imgs; init=7)
```
for enhanced readability

TWiedemann replied:
Done

lgoettgens commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+"""
+function map_word(w::WeylGroupElem, genimgs::Vector; init=nothing)
+  @req length(genimgs) == number_of_generators(parent(w)) "Length of vector of images does not equal rank of Weyl group"
+  return Oscar.map_word([Int(i) for i in word(w)], genimgs, init=init)
```
```suggestion
  return map_word(Int.(word(w)), genimgs; init=init)
```
this should help a bit with type stability for the empty word

TWiedemann replied:
Done

lgoettgens commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+Returns two objects: The subgroup `U` of `W` generated by `[inv(w)*u*w for u in gens(W)[vec]]`, whose generators are set to be this vector and which is itself a `WeylGroup`, and the embedding of `U` into `W`. The elements of `vec` must be pairwise distinct integers in `1:number_of_generators(W)` and `vec` must be non-empty.
+"""
+function parabolic_subgroup(W::WeylGroup, vec::Vector{<:Integer}, w::WeylGroupElem=one(W))
+  @req length(unique(vec)) == length(vec) "Elements of vector are not pairwise distinct"
```
```suggestion
  @req allunique(vec) "Elements of vector are not pairwise distinct"
  @req all(i -> 1 <= i <= number_of_generators(W), vec) "Invalid indices"
```
better to check that as well, should be cheap

TWiedemann replied:
Done

lgoettgens commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+end
+
+@doc raw"""
+    parabolic_subgroup(W::WeylGroup, vec::Vector{<:Integer}, w::WeylGroupElem) -> WeylGroup, Map{WeylGroup, WeylGroup}
```
```suggestion
    parabolic_subgroup(W::WeylGroup, vec::Vector{<:Integer}, w::WeylGroupElem=one(W)) -> WeylGroup, Map{WeylGroup, WeylGroup}
```
add the default value here as well

TWiedemann replied:
Done

lgoettgens commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+    for i in vec
+      for j in setdiff(1:number_of_generators(W), vec)
+        @req coxeter_matrix_entry_from_cartan_matrix(cm, i, j) == 2 "Input vector does not describe a factor of the Weyl group, so there is no projection homomorphism"
+      end;
```
```suggestion
      end
```

lgoettgens commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+    cm = cartan_matrix(root_system(W))
+    for i in vec
+      for j in setdiff(1:number_of_generators(W), vec)
+        @req coxeter_matrix_entry_from_cartan_matrix(cm, i, j) == 2 "Input vector does not describe a factor of the Weyl group, so there is no projection homomorphism"
```
```suggestion
        @req is_zero_entry(cm, i, j) "Input vector does not describe a factor of the Weyl group, so there is no projection homomorphism"
```
Is this mathematically the same? (I am not sure atm.) If yes, this would be a bit more performant

TWiedemann replied:
I'm pretty sure it's mathematically the same (`cm(i, j) == 0` iff the simple roots with indices `i` and `j` are orthogonal iff $s_i$ commutes with $s_j$), so I made this change. Also wrapped the line here.

TWiedemann committed:
Fix small issues

TWiedemann commented:
@thofma @fingolfin @lgoettgens Thanks for the help and suggestions! I fixed everything.

One non-nightly test fails: Am I not allowed to spread a string over several lines with "\", as in WeylGroup.jl:398?
```
        @req is_zero_entry(cm, i, j) begin
          "Input vector does not describe a factor of the Weyl group, \
          so there is no projection homomorphism"
        end
```
[This was apparently introduced in Julia 1.7](https://discourse.julialang.org/t/single-string-on-multiple-lines-without-n/18099/7), which might be the problem.

lgoettgens commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,159 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+      for j in setdiff(1:number_of_generators(W), vec)
+        @req is_zero_entry(cm, i, j) begin
+          "Input vector does not describe a factor of the Weyl group, \
+          so there is no projection homomorphism"
```
```suggestion
          "Input vector does not describe a factor of the Weyl group, so there is no projection homomorphism"
```
does it work like this? I know that the line is then technically overly long, but that would be fine here IMO

lgoettgens commented:
Could you please also use `map_word` in the `isomorphism` functions in the same file (cf. https://github.com/oscar-system/Oscar.jl/pull/4478#discussion_r1922579409)? This would then provide as another indirect test of the newly added function

fingolfin committed:
Apply suggestions from code review

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

TWiedemann committed:
Use map_word in isomorphism(PermGroup, ...)

TWiedemann committed:
Add ref to map_word(::WeylGroupElem) in map_word(::FPGroupElem) and map_word(::PcGroupElem)

TWiedemann commented:
> Could you please also use `map_word` in the `isomorphism` functions in the same file (cf. [#4478 (comment)](https://github.com/oscar-system/Oscar.jl/pull/4478#discussion_r1922579409))? This would then provide as another indirect test of the newly added function

Done. Also added links to the new `map_word` in the docs of the old `map_word` functions.

I have removed the draft status.

lgoettgens reviewed (approved):
Looks good, thanks! @fingolfin could you please have a look again as well?

fingolfin commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,158 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+    for i in vec
+      for j in setdiff(1:number_of_generators(W), vec)
+        @req is_zero_entry(cm, i, j) begin
+          "Input vector does not describe a factor of the Weyl group, so there is no projection homomorphism"
```
```suggestion
          "Input vector must describe a direct factor of the Weyl group"
```

fingolfin commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,158 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+    # In other words, vec describes a union of irreducible components of the Coxeter diagram.
+    cm = cartan_matrix(root_system(W))
+    for i in vec
+      for j in setdiff(1:number_of_generators(W), vec)
```
This recomputes `setdiff` on each iteration, can be avoided by switching the loops:
```suggestion
    for i in setdiff(1:number_of_generators(W), vec)
      for j in vec
```

fingolfin commented on experimental/LieAlgebras/src/WeylGroup.jl:
````diff
@@ -257,3 +257,158 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+```
+"""
+function parabolic_subgroup_with_projection(
+  W::WeylGroup, vec::Vector{<:Integer}; check_factor::Bool=true
````
We often have optional checks and they almost always are controlled by a kwarg `check`. Perhaps stick with that here, too?
```suggestion
  W::WeylGroup, vec::Vector{<:Integer}; check::Bool=true
```

TWiedemann replied:
Makes sense

fingolfin commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,158 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+Here `P, emb = `[`parabolic_subgroup`](@ref)`(W, vec)`
+and `proj` is the projection map from `W` onto `P`,
+which is a left-inverse of `emb`.
+If `P` is not a factor of `W`, an error occurs.
```
Maybe document the optional kwarg `check_factor` / `check`? Or is it deliberately undocumented?

TWiedemann replied:
It was a deliberate choice to not document because it seemed to me that the user should not normally disable the check. I had the `set_properties` argument of `isomorphism` in mind ([previous disucssion](https://github.com/oscar-system/Oscar.jl/pull/4478#discussion_r1920039870)); however, I do see now that the situations are rather different: `set_properties = false` really only makes sense for test purposes, which is not the case for `check` here. So yes, I agree that this should be documented. Will do that.

fingolfin commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,158 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+  @req all(i -> 1 <= i <= number_of_generators(W), vec) "Invalid indices"
+  cm = cartan_matrix(root_system(W))[vec, vec]
+  para = weyl_group(cm)
+  genimgs = map(x -> inv(w) * x * w, gens(W)[vec])
```
```suggestion
  genimgs = [conj(W[i], w) for i in vec]
```

fingolfin commented on experimental/LieAlgebras/test/WeylGroup-test.jl:
```diff
@@ -135,4 +135,99 @@
+        # On subsequent runs, conjugate by random elements
+        conj = rand(W)
+        para, emb = parabolic_subgroup(W, vec, conj)
+        genimgs = [inv(conj) * gen(W, vec[i]) * conj for i in 1:length(vec)]
```
```suggestion
        r = rand(W)
        para, emb = parabolic_subgroup(W, vec, r)
        genimgs = [conj(W[i], r) for i in vec]
```

fingolfin commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,158 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+  factor, emb = parabolic_subgroup(W, vec)
+  # Generators of W are mapped to the corresponding generators of factor,
+  # or to 1 if there is no corresponding generator
+  proj_gen_imgs = [one(factor) for _ in 1:number_of_generators(W)]
```
```suggestion
  proj_gen_imgs = fill(one(factor), number_of_generators(W))
```
or (my personal preference but it's a matter of taste)
```suggestion
  proj_gen_imgs = fill(one(factor), ngens(W))
```

lgoettgens replied:
Note that there is a semantical difference here. `fill` uses the same object for all positions, while the list comprehension uses different (but identical) objects.

So the former is slightly more performant, but may not be used in contexts with mutating functions. But as far as I can see that is not the case here

TWiedemann replied:
Ah, good. I had wondered whether there was a better way to initialise the vector, thanks.

fingolfin commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,158 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+See also: [`map_word(::Union{FPGroupElem, SubFPGroupElem}, ::Vector)`](@ref),
+[`map_word(::Union{PcGroupElem, SubPcGroupElem}, ::Vector)`](@ref).
+Note that `map_word(::WeylGroupElem)` accepts but ignores the `genimgs_inv` keyword argument
+because the generators of a Weyl group are always self-inverse.
```
```suggestion
Note that `map_word(::WeylGroupElem)` accepts the `genimgs_inv` keyword argument
for consistency with other `map_word` methods, but ignores it because the
generators of a Weyl group are always self-inverse.
```

fingolfin reviewed (commented):
Looks good to me in principle, but I have a few small suggestions.

TWiedemann committed:
Adapt error message in parabolic_subgroup_with_projection

Co-authored-by: Max Horn <max@quendi.de>

TWiedemann committed:
Use conj

Co-authored-by: Max Horn <max@quendi.de>

TWiedemann committed:
Use conj in parabolic subgroup test

Co-authored-by: Max Horn <max@quendi.de>

TWiedemann committed:
Clarify doc of map_word

Co-authored-by: Max Horn <max@quendi.de>

TWiedemann committed:
Use fill for vector initialisation

Co-authored-by: Max Horn <max@quendi.de>

TWiedemann committed:
Switch for loops

Co-authored-by: Max Horn <max@quendi.de>

TWiedemann commented:
@fingolfin Thanks! I will implement the remaining suggestions tomorrow.

TWiedemann committed:
Document check

TWiedemann commented:
Fixed the last two issues (`check_factor` -> `check` and added documentation for `check`).

lgoettgens commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,160 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+end
+
+@doc raw"""
+    parabolic_subgroup_with_projection(W::WeylGroup, vec::Vector{<:Integer}) -> WeylGroup, Map{WeylGroup, WeylGroup}, Map{WeylGroup, WeylGroup}
```
```suggestion
    parabolic_subgroup_with_projection(W::WeylGroup, vec::Vector{<:Integer}; check::Bool=true) -> WeylGroup, Map{WeylGroup, WeylGroup}, Map{WeylGroup, WeylGroup}
```

TWiedemann replied:
Thanks, I just realised myself during a seminar that I forgot about this :D

lgoettgens commented on experimental/LieAlgebras/src/WeylGroup.jl:
```diff
@@ -257,3 +257,160 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
+Here `P, emb = `[`parabolic_subgroup`](@ref)`(W, vec)`
+and `proj` is the projection map from `W` onto `P`,
+which is a left-inverse of `emb`.
+If `check = true`, then it is checked whether `vec` actually describes
```
```suggestion

If `check = true`, then it is checked whether `vec` actually describes
```
new paragraph for better readability.

TWiedemann committed:
Update experimental/LieAlgebras/src/WeylGroup.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

TWiedemann committed:
Update experimental/LieAlgebras/src/WeylGroup.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>
